src/main/webapp/app/exam/manage/exam-status.component.ts

```diff
@@ -0,0 +1,187 @@
+    }
+
+    /**
+     * Auxiliary method indicating whether all steps of Exam preparation are done
+     * @private
+     */
+    private isExamPreparationFinished(): boolean {
+        return this.examConfigurationSteps.every((suffix) => this.configurationStepToFlagMap.get(suffix));
+    }
+
+    /**
+     * Sets the conductionState according to the current situation
+     * @private
+     */
+    private setConductionState(): void {
+        if (this.examAlreadyEnded()) {
+            this.examConductionState = ExamConductionState.FINISHED;
+        } else if (this.examAlreadyStarted()) {
+            this.examConductionState = ExamConductionState.RUNNING;
+        } else {
+            this.examConductionState = ExamConductionState.PLANNED;
+        }
+    }
+
+    /**
+     * Sets the reviewState according to the current situation
+     * @private
+     */
+    private setReviewState(): void {
+        if (!this.exam.examStudentReviewEnd) {
+            this.examReviewState = ExamReviewState.UNSET;
+        } else if (this.isExamReviewRunning()) {
+            this.examReviewState = ExamReviewState.RUNNING;
+        } else if (this.exam.examStudentReviewEnd.isBefore(dayjs())) {
+            this.examReviewState = ExamReviewState.FINISHED;
+        } else {
+            this.examReviewState = ExamReviewState.PLANNED;
+        }
+    }
+
+    /**
+     * Auxiliary method that determines the state of the different sub steps of exam preparation and stores them in a map
+     * Finally determines whether every sub step is sufficiently fulfilled and therefore exam preparation is finished
+     * @private
+     */
+    private setExamPreparation(): void {
+        // Step 1.1: Configure exercises
+        this.configurationStepToFlagMap.set(ExamConfigurationStep.CONFIGURE_EXERCISES, this.areAllExercisesConfigured());
+        // Step 1.2: Register students
+        this.configurationStepToFlagMap.set(ExamConfigurationStep.REGISTER_STUDENTS, this.examChecklistService.checkAtLeastOneRegisteredStudent(this.exam));
+        // Step 1.3: Generate student exams
+        this.configurationStepToFlagMap.set(ExamConfigurationStep.GENERATE_STUDENT_EXAMS, this.examChecklistService.checkAllExamsGenerated(this.exam, this.examChecklist));
+        // Step 1.4: Prepare exercise start
+        const allExamsGenerated = this.configurationStepToFlagMap.get(ExamConfigurationStep.GENERATE_STUDENT_EXAMS)!;
+        this.configurationStepToFlagMap.set(ExamConfigurationStep.PREPARE_EXERCISE_START, !!this.examChecklist.allExamExercisesAllStudentsPrepared && allExamsGenerated);
+        this.examPreparationFinished = this.isExamPreparationFinished();
+    }
+
+    /**
+     * Indicates whether the exam already started
+     * @private
+     */
+    private examAlreadyStarted(): boolean {
+        return this.exam.startDate! && this.exam.startDate!.isBefore(dayjs());
```

Comment:
```suggestion
        return this.exam.startDate && this.exam.startDate.isBefore(dayjs());
```
The check if the `startDate` is defined shouldn’t require the `!` on the left side. On the right side it should not be needed because the left side already checked that is defined.

Reply:
This was my perception as well. Unfortunately, the compiler has other ideas. 😅 I would prefer omitting `undefined` as return type. What do you think?
![image](https://user-images.githubusercontent.com/80622272/152150750-1fae1b0d-1cc3-499e-87db-1853b9cc4dac.png)

Reply:
Interesting, well if the compiler wants to have the `!` then so they should definitely stay to avoid the `| undefined` as return type.

(Would `this.exam.startDate?.isBefore(dayjs()) ?? false` work? I’m not sure if that would improve the code or if the current version is better.)

Reply:
Good catch, this one would work. But I share your concerns that this might make the code a bit more complicated than it is now.
In any case, I think it is sufficient if we keep the exclamation sign for the first `startDate` and remove the second one.

Reply:
I removed the exclamation marks in 93c3171